fabric-api-base/src/testmodClient/java/net/fabricmc/fabric/test/base/client/ThreadingImpl.java

```diff
@@ -0,0 +1,203 @@
+ * task will release the test thread semaphore and re-block on its own semaphore again and the cycle continues. When the
+ * test phase is over (i.e. when the test thread wants to wait a tick), the client and server semaphores will be
+ * released while leaving {@linkplain #taskToRun} as {@code null}, which they will interpret to mean they are to
+ * continue into {@linkplain #PHASE_TICK}.
+ *
+ * <p>The reason these phases were chosen are to make client-server interaction in singleplayer as consistent as
+ * possible. The task queues are when most packets are handled, and without them being run in sequence it would be
+ * unspecified whether a packet would be handled on the current tick until the next one. The server task queue is before
+ * the client so that changes on the server appear on the client more readily. The test phase is run after the task
+ * queues rather than at the end of the physical tick (i.e. {@code MinecraftClient}'s and {@code MinecraftServer}'s
+ * {@code tick} methods), for no particular reason other than to avoid needing a 5th phase, and having a power of 2
+ * number of phases is convenient when using {@linkplain Phaser}, as it doesn't break when the phase counter overflows.
+ *
+ * <p>Other challenges include that a client or server can be started during {@linkplain #PHASE_TEST} but haven't
+ * reached their semaphore code yet meaning they are unable to accept tasks. This is solved by setting a flag to true
+ * when the client/server is ready to accept tasks. Also the client will block on the integrated server starting and
+ * stopping. This is solved by first deferring those operations until {@linkplain #PHASE_TICK} if they are being run
+ * inside a test phase task (which is a minor difference from vanilla), and then ensuring the client is still running
+ * the phase logic and is able to accept tasks while it is waiting for the server.
+ */
+public final class ThreadingImpl {
+	private ThreadingImpl() {
+	}
+
+	public static final int PHASE_TICK = 0;
+	public static final int PHASE_SERVER_TASKS = 1;
+	public static final int PHASE_CLIENT_TASKS = 2;
+	public static final int PHASE_TEST = 3;
+	private static final int PHASE_MASK = 3;
+
+	public static final Phaser PHASER = new Phaser();
+
+	public static volatile boolean isClientRunning = false;
+	public static volatile boolean clientCanAcceptTasks = false;
+	public static final Semaphore CLIENT_SEMAPHORE = new Semaphore(0);
+
+	public static volatile boolean isServerRunning = false;
+	public static volatile boolean serverCanAcceptTasks = false;
+	public static final Semaphore SERVER_SEMAPHORE = new Semaphore(0);
+
+	@Nullable
+	public static Thread testThread = null;
+	public static final Semaphore TEST_SEMAPHORE = new Semaphore(0);
+
+	@Nullable
+	public static Runnable taskToRun = null;
+
+	public static void enterPhase(int phase) {
+		while ((PHASER.getPhase() & PHASE_MASK) != phase) {
```

Comment:
This is the only bit that is midly confusing to me (I think), is this method blocking until the current phase has been reached from somewhere else? My tottal lack of knowledge of Phaser doesnt help

Reply:
This method blocks until the given tick phase has been reached on all threads registered to the `Phaser` ("parties"), while allowing other parties to advance through the other tick phases (which is what the loop is for). The `arriveAndAwaitAdvance` is a barrier that all parties must call before they are all unblocked at once.

A `Phaser` maintains a counter called the "phase" which increments each time a barrier is crossed. It starts at 0 and when it surpasses `Integer.MAX_VALUE` it resets back to 0 again (negative values happen when the `Phaser` is terminated, which is not functionality we're using). Using the lower 2 bits of this counter that already exists tells us which of our 4 tick phases we're currently in. It's possible to track that separately by subclassing the `Phaser` and overriding its `onAdvance` method, but I figured why not take advantage of a counter that already exists.